Sources/generate-swift-syntax-builder/gyb_syntax_support/StmtNodes.py

```diff
@@ -242,6 +243,16 @@
              Child('RightParen', kind='RightParenToken'),
          ]),
 
+    # has-symbol-condition -> '#_hasSymbol' '(' expression ')'
+    Node('HasSymbolCondition', name_for_diagnostics="'#_hasSymbol' condition",
+         kind='Syntax',
+         children=[
+             Child('HasSymbolKeyword', kind='HasSymbolToken'),
+             Child('LeftParen', kind='LeftParenToken'),
+             Child('Expression', kind='Expr'),
```

Comment:
I know that it’s also modeled like this in C++ and I don’t know exactly how `#_hasSymbol` works but is expression the right way to model the symbol? Wouldn’t `QualifiedDeclName` be better?

Reply:
I'm very much not sure about the best way to approach this, but my impression was that `QualifiedDeclName` is more limited than I need. I need to be able to unambiguously identify functions, methods, properties accessors, nominal types, Clang functions and global vars, and maybe even protocol conformances; basically any declaration that can be referenced in an expression. It seems like a way to accomplish this is to parse expressions and then accept a subset of them (post type checking) that can be recognized as identifying a single declaration. I recognize that this is a bit of an unusual approach. The advantages I see are that I can leverage existing type checking logic to do the declaration lookup in context. This is especially helpful when it comes to disambiguation since I can use coercion expressions to disambiguate overloaded functions and methods.

What do you think, given the requirements?

Reply:
Ah, I see. I think coercion is the feature that requires you to use an expression here. `QualifiedDeclName` wouldn’t be able to distinguish those AFAIK.

Do you think adding a test case for such a use case that requires expression parsing would make sense?

Reply:
Yup, added one.